Bitpack can crash if empty slice is provided to UnpackInto()

Turns out that it's too easy to crash `bitpack` by handing over an empty slice:

```
ts=2024-05-30T13:15:57.990+02:00 level=warn msg="Incorrect number of entries in variable block size file. Expected file length -12, have 4" version=devel block=110 column=sip
panic: runtime error: integer divide by zero
goroutine 48 [running]:
github.com/fako1024/gotools/bitpack.UnpackInto({0xc000314000?, 0xc000de7250?, 0x4000?}, {0xc0003644c0?, 0x0?, 0x0?})
 	/home/rpm/go/pkg/mod/github.com/fako1024/gotools/bitpack@v0.0.0-20240419120819-e06ca9ce9e88/bitpack.go:43 +0x498
 github.com/els0r/goProbe/pkg/goDB.(*DBWorkManager).readBlocksAndEvaluate(0xc00034e360, 0xc0000d1b80, {0xf9a8a0, 0xc000d80130}, 0xc00052feb8)
 	/usr/src/packages/src/osco/observability/pkg/OSAGtmi/goProbe/pkg/goDB/DBWorkManager.go:641 +0x129f
 github.com/els0r/goProbe/pkg/goDB.(*DBWorkManager).grabAndProcessWorkload.func1()
 	/usr/src/packages/src/osco/observability/pkg/OSAGtmi/goProbe/pkg/goDB/DBWorkManager.go:496 +0x545
 created by github.com/els0r/goProbe/pkg/goDB.(*DBWorkManager).grabAndProcessWorkload in goroutine 1
: 	/usr/src/packages/src/osco/observability/pkg/OSAGtmi/goProbe/pkg/goDB/DBWorkManager.go:455 +0xbd
```

DoD

- [x] Add test
- [x] Add actual guard (at least prevent the panic) or a second method that is safe

--- a/bitpack/bitpack.go
+++ b/bitpack/bitpack.go
@@ -39,7 +39,17 @@ func Pack(data []uint64) []byte {
 // UnpackInto decompresses a compressed byte slice into a pre-existing slice of
 // uint64 values (which will be allocated / grown in case its capacity is insufficient)
 func UnpackInto(b []byte, res []uint64) []uint64 {
+
+	// If the byte slice is empty, truncate and return the buffer
+	if len(b) == 0 {
+		return res[:0]
+	}
+
+	// If the number of unpacked bytes is zero, truncate and return the buffer
 	neededBytes := int(b[0])
+	if neededBytes == 0 {
+		return res[:0]
+	}
 	nElements := (len(b) - 1) / neededBytes
 
 	if cap(res) < nElements {
@@ -73,7 +83,17 @@ func UnpackInto(b []byte, res []uint64) []uint64 {
 // Unpack decompresses a previously compressed data slice into the original slice of
 // uint64 values
 func Unpack(b []byte) []uint64 {
+
+	// If the byte slice is empty, return an empty result
+	if len(b) == 0 {
+		return []uint64{}
+	}
+
+	// If the number of unpacked bytes is zero, return an empty result
 	neededBytes := int(b[0])
+	if neededBytes == 0 {
+		return []uint64{}
+	}
 	nElements := (len(b) - 1) / neededBytes
 	res := make([]uint64, nElements)
 

--- a/bitpack/bitpack_test.go
+++ b/bitpack/bitpack_test.go
@@ -60,6 +60,18 @@ func TestTable(t *testing.T) {
 
 func TestCorruptInput(t *testing.T) {
 	require.Zero(t, Len([]byte{0x0}))
+	require.Zero(t, Len([]byte{}))
+	require.Zero(t, Len(nil))
+
+	buf, res := []byte{0}, make([]uint64, 64)
+	require.Empty(t, UnpackInto(nil, res))
+	require.Empty(t, UnpackInto([]byte{}, res))
+	require.Empty(t, UnpackInto(buf, res))
+	require.Empty(t, UnpackInto(buf, nil))
+
+	require.Empty(t, Unpack(nil))
+	require.Empty(t, Unpack([]byte{}))
+	require.Empty(t, Unpack(buf))
 }
 
 func TestAllByteWidths(t *testing.T) {

